tests/scripts/git-history

```diff
@@ -56,31 +56,32 @@ while read -r rev; do
     git checkout -q "${rev}"
 
     # Compute the total number of changes (line insertions + line deletions) in
-    # the commit.
+    # all the files in the commit.
     changes=$(git show --pretty="" --numstat | awk '{change+=$1+$2} END {print change}')
 
-    # If the above is more than 50K, print an error
+    # If the total number of changes is more than 50K, print an error
     if [ ${changes} -gt 50000 ]
     then
         echo "Commit ${rev} is unusually large: it changes ${changes} lines" 1>&2
     fi
 
-    # Find all non-empty binary files in the current commit (use the output of
-    # the "file" command, excluding text files and PNG images)
+    # Loop over the files modified in the current commit.
     for fname in $(git show --pretty="" --name-only ${rev})
     do
         if [ -e $fname ] # exclude deleted files
         then
+            # Report binary files (non-text files as reported by "file")
             binary="$(file -b ${fname} | grep -v 'PNG image' | grep -v 'text')"
-            if [ -n "${binary}" ]
+            if [ -n "${binary}" ] # non-empty files
             then
                 echo "Commit ${rev} contains a binary file: ${fname} (${binary})" 1>&2
             fi
 
+            # Report files greater than 50K
             size="$(du -k ${fname} | cut -f1)"
-            if [ ${size} -gt 400 ] # currently no file larger than 400K in mfem
+            if [ ${size} -gt 50 ]
```

Comment:
Simply modifying an existing large file will trigger the error. 50K is too small, since many source files are concerned.
I suggest to create an issue with the improvements we want:
- restricting size check to new files?
- checking for large number of commit history (re-writing history)

Reply:
Yes, it is not clear what to do here exactly. I'd like to think about it a bit more before we merge.

The problem is that if a check fails, the second-stage Travis tests will not run, and so we shouldn't probably merge its PR. Allowing checks to optionally fail is also not acceptable because we can miss a legitimate problem.

Therefore either the checks should never be OK to fail or we should filter the exceptions.

So what do we do about the "large number of changes" check?

Reply:
Maybe we can add a third stage and move there the checks that we can ignore when they fail?

Reply:
It seems to me that we try to address "catastrophic" situations, so we may accept not to capture everything but only clearly wrong situations. That’s for the failing jobs.

We could have a third stage of "acceptable failures" with to goal to perform similar checks with lower bounds.

When it comes to legitimate massive changes, we could either use filtering, or _tags in commits messages_. For example, we could skip checking for "large number of changes" if the commit message contains [REFORMAT], or skip "png files addition" if [PNG]... The main drawback being that to add these tags _a posteriori_ one would have to rewrite the branch history.

Such a tagging mechanism could also be implemented at the PR level (tags in PR title or message). Preventing from history rewriting, but requiring more attention.